dds_web/api/user.py

```diff
@@ -982,6 +984,79 @@ def get(self):
         return {"token": update_token_with_mfa(token_claims)}
 
 
+class RequestTOTPActivation(flask_restful.Resource):
+    """Request to switch from HOTP to TOTP for second factor authentication."""
+
+    @auth.login_required
+    def put(self):
+
+        user = auth.current_user()
+        json_info = flask.request.json
+        activate_totp = json_info.get("activate_totp", False)
+        if activate_totp:
+            if user.totp_enabled:
+                return {
+                    "message": "Nothing to do, two-factor authentication with app is already enabled for this user."
+                }
+
+            # Not really necessary to encrypt this
+            token = encrypted_jwt_token(
+                username=user.username,
+                sensitive_content=None,
+                expires_in=datetime.timedelta(
+                    seconds=3600,
+                ),
+                additional_claims={"act": "totp"},  # Open for suggestions
+            )
+
+            link = flask.url_for("auth_blueprint.activate_totp", token=token, _external=True)
+            # Send activation token to email to work as a validation step
+            # TODO: refactor this since the email sending code is replicated in many places
+            recipients = [user.primary_email]
+
+            # Fill in email subject with sentence subject
+            subject = f"Request to activate two-factor with authenticator app for SciLifeLab Data Delivery System"
+
+            msg = flask_mail.Message(
+                subject,
+                recipients=recipients,
+            )
+
+            # Need to attach the image to be able to use it
+            msg.attach(
+                "scilifelab_logo.png",
+                "image/png",
+                open(
+                    os.path.join(flask.current_app.static_folder, "img/scilifelab_logo.png"), "rb"
+                ).read(),
+                "inline",
+                headers=[
+                    ["Content-ID", "<Logo>"],
+                ],
+            )
+
+            msg.body = flask.render_template(
+                f"mail/request_activate_totp.txt",
+                link=link,
+            )
+            msg.html = flask.render_template(
+                f"mail/request_activate_totp.html",
+                link=link,
+            )
+
+            AddUser.send_email_with_retry(msg)
+            return {
+                "message": "Please check your email and follow the attached link to activate two-factor with authenticator app."
+            }
+        else:
+            if not user.totp_enabled:
```

Comment:
This is just a personal preference but it would make more sense to me to have it like this instead

```
if user.totp_enabled:
    user.deactivate_totp()
    return ...

return ....
```